### ISSUE TITLE
kubectl describe node shows the Allocatable memory  in milli bytes ('m')

### ISSUE THREAD

**Issue by kmadnani** [kubernetes/kubernetes#74986]
**What happened**:
When i do a kubectl describe node, it shows me Allocatable Memory with a unit of 'm' which is usually used for cpu (millicores). The k8s docs don't mention this unit as one of the standard ones for specifying memory. 

Allocatable:
 cpu:                7800m
 ephemeral-storage:  86069648Ki
 hugepages-1Gi:      0
 hugepages-2Mi:      0
 memory:             **15961844721600m**
 pods:               110

**What you expected to happen**:
The memory should be mentioned in one of the standard units as mentioned in k8s docs:
`E, P, T, G, M, K, Ei, Pi, Ti, Gi, Mi, Ki.`

**How to reproduce it (as minimally and precisely as possible)**:
kubectl describe node <node-name>

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`): v1.12.5
- Cloud provider or hardware configuration: vmware
- OS (e.g: `cat /etc/os-release`): Ubuntu 16.04.4 LTS
- Kernel (e.g. `uname -a`): Linux us06000c3-master-k8-03 4.4.0-130-generic #156-Ubuntu SMP Thu Jun 14 08:53:28 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
- Install tools: Kubespray
- Others:

**Comment by kmadnani** [comment]
@kubernetes/sig-Node-bugs

**Comment by liucimin** [comment]
Are you sure the describe message is this?
Or should u show us the node's true memory?Not from the k8s.
Or try the lastest version.

**Comment by kmadnani** [comment]
@liucimin  Yes i am sure the describe message is this. Infact other nodes are showing the same behavior too.
This is from other node:
```
Allocatable:
 cpu:                3900m
 ephemeral-storage:  86069648Ki
 hugepages-2Mi:      0
 memory:             13206125965600m
 pods:               110
```
```
$ free
              total        used        free      shared  buff/cache   available
Mem:       16432380      863852      948944      146288    14619584    14934864
Swap:             0           0           0
```

We use kubespray, in which the latest stable version is v1.12.5 for kubespray release v2.8.3

**Comment by perrywang** [comment]
same on 1.12

**Comment by zonArt** [comment]
same on 1.9

**Comment by apjneeraj** [comment]
Same here too in k8s 1.12(kops cluster). But interestingly it happened after I enforced reservations for memory/cpu. Earlier it was showing correct units 
> enforceNodeAllocatable: "pods,kube-reserved" 

I used following snipped from here to apply the limits: 
https://github.com/kubernetes/kops/issues/3762#issuecomment-356478668

```
Capacity:
 attachable-volumes-aws-ebs:  25
 cpu:                         2
 ephemeral-storage:           64989928Ki
 hugepages-1Gi:               0
 hugepages-2Mi:               0
 memory:                      7872332Ki
 pods:                        110
Allocatable:
 attachable-volumes-aws-ebs:  25
 cpu:                         1250m
 ephemeral-storage:           59894717546
 hugepages-1Gi:               0
 hugepages-2Mi:               0
 memory:                      5878971596800m
 pods:                        110
```

**Comment by apjneeraj** [comment]
So, the issue does not appear if I reduce the set limit from 
```
evictionHard: memory.available<750Mi,nodefs.available<10%,nodefs.inodesFree<5%,imagefs.available<10%,imagefs.inodesFree<5%
```
to 
```
evictionHard: memory.available<500Mi,nodefs.available<10%,nodefs.inodesFree<5%,imagefs.available<10%,imagefs.inodesFree<5%
```
Also, reduced the limit for
```
    kubeReserved:
      cpu: 750m
      memory: 1.3Gi
```
I think it then does not convert it millibytes.

**Comment by 2rs2ts** [comment]
So is the problem that if you reserve too many resources Kubernetes uses a bad unit? Or is it caused by using % based eviction thresholds?

We are seeing not millibytes but just bytes, but it's messing up [`kubectl-view-utilization`]'s parsing https://github.com/etopeter/kubectl-view-utilization

**Comment by lllamnyp** [comment]
I do not think this is a problem.
In a comment above we see
```
kubeReserved:
      cpu: 750m
      memory: 1.3Gi
```
Now 1.3Gi is a non-integer amount of bytes, specifically `2 ** 30 * 13 / 10`.  If you subtract this from the overall capacity, you will have also a non-integer amount of bytes allocatable. So long as kubelet isn't trying anything funny with cgroups, this should be a non-issue.

**Comment by thst71** [comment]
I see the same issue, I do not understand why borked display values are a non-issue. YES, it may be the correct value, but millibytes can hardly be seen as a standard way of expressing memory settings in a Ti world.

**Comment by lllamnyp** [comment]
@thst71 I think it would be too much to require a general-purpose library that is designed to represent integer and non-integer quantities alike to have context-awareness of the nature of the quantity it is trying to represent. I mean, sure, the developers of k8s could create a new struct such as `MemoryQuantity` that implements the `Quantity` interface, and then update the k8s API, but that seems like breaking backwards compatibility for something (that seems to me as) minor. As Rob Pike (IIRC) puts it, the importance of an issue must be demonstrated prior to fixing it.

**Comment by invsblduck** [comment]
IMO, this breaks [POLA](https://en.wikipedia.org/wiki/Principle_of_least_astonishment) pretty bad. I've worked on computers 30 years and had to scratch my head when I saw this (like, "wtf?!"). I don't care enough to fix it myself, so I'm certainly not whining that anyone else should fix it&#8212;just thought I'd +1 the sentiment while I was already grumbling about doing it wrong and scraping `kubectl describe` stdout in the first place. :smile:
